### PR TITLE
GH-36789: [C++] Support divide(duration, duration)

### LINKED ISSUE #36789: [C++][Python] Add division kernel: duration / duration -> float
### Describe the enhancement requested

Numpy allows division of timedelta64 by timedelta64 types, resulting in plain floating points (units cancel!)

```python
import pyarrow as pa
import numpy as np

one_second_numpy = np.timedelta64(1, "s")
one_second_arrow = pa.scalar(1, pa.duration("s"))

assert one_second_numpy / one_second_numpy == 1.0  # ✔
assert one_second_arrow / one_second_arrow == 1.0  # TypeError: unsupported operand
```

### Component(s)

Python

### PR DESCRIPTION
**Pull request by js8544**
### Rationale for this change

Support divide(duration, duration), as pandas and numpy already do.

### What changes are included in this PR?

Add kernels divide(duration, duration)->float64 and divide_checked(duration, duration)->float64

### Are these changes tested?

Yes

### Are there any user-facing changes?

No

* Closes: #36789

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 1]
```diff
@@ -1695,6 +1695,7 @@ TEST_F(ScalarTemporalTest, TestTemporalMultiplyDuration) {
 }
 
 TEST_F(ScalarTemporalTest, TestTemporalDivideDuration) {
+  // div(duration, integer)
```
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1698]: ... -> integer

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 2]
```diff
@@ -1718,6 +1719,13 @@ TEST_F(ScalarTemporalTest, TestTemporalDivideDuration) {
                                       CallFunction("divide_checked", {durations, zeros}));
     }
   }
+
+  // div(duration, duration) -> float64
+  auto left = ArrayFromJSON(duration(TimeUnit::SECOND), "[1, 2, 3, 4]");
+  auto right = ArrayFromJSON(duration(TimeUnit::MILLI), "[4000, 300, 20, 1]");
+  auto expected = ArrayFromJSON(float64(), "[0.25, 6.666666666666667, 150, 4000]");
+  CheckScalarBinary("divide", left, right, expected);
+  CheckScalarBinary("divide_checked", left, right, expected);
```
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1728]: Should probably add division of MILLIs by SECOND to ensure we get unit handling correctly on the other direction as well.

**Comment by felipecrv** [comment]
This implementation is very elegant and simple, but it's giving up some accuracy we can reclaim with a bit more careful handling of the units.

If I understand correctly each DURATION operand is being converted to a FLOAT64 before the division, so `x NANO / y MILLI` will first divide `x` by `1e9`, then `y` by `1e3`, and then perform another division with those two results.

There is a considerable accuracy loss for low-valued `x` [1]. An alternative way to do it is to first build a fraction from the units. In this `NANO/MILLI` example that would be `1e-9/1e-3 = 1e-6`.

So the actual computation would be `(x * 1e-6) / y`. This is also more efficient: one `*` and one `/` instead of 3 `/`.

My concern can probably be ignored if numpy/pandas does the naive scaling on both operands and the divide again.

@pitrou what do you think.

image

[1] https://herbie.uwplse.org/demo/8908317b92cc5eb8646c2968ab4956e4e96c9cea.2.0/graph.html

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 1 above)
Comment (js8544) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1698]: fixed

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 2 above)
Comment (js8544) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc]: Another test added.

**Comment by js8544** [comment]
> If I understand correctly each DURATION operand is being converted to a FLOAT64 before the division, so `x NANO / y MILLI` will first divide `x` by `1e9`, then `y` by `1e3`, and then perform another division with those two results.

Not really. [`ArithmeticFunction`](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L638) will cast them to the finer common unit, not SECOND. So in the case of `x NANO / y MILLI`, `y` would be casted to NANO first, resulting in `x / (y * 1e6)`. I believe this is effectively the same as the algorithm you proposed.

**Review by felipecrv — APPROVED** [review]

**Comment by felipecrv** [comment]
> > If I understand correctly each DURATION operand is being converted to a FLOAT64 before the division, so `x NANO / y MILLI` will first divide `x` by `1e9`, then `y` by `1e3`, and then perform another division with those two results.
> 
> Not really. [`ArithmeticFunction`](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L638) will cast them to the finer common unit, not SECOND. So in the case of `x NANO / y MILLI`, `y` would be casted to NANO first, resulting in `x / (y * 1e6)`. I believe this is effectively the same as the algorithm you proposed.
> 
> I also added a DispatchBest test to demonstrate this.

Thanks. TIL. Having the dispatching code find the best unit is equivalent to what I was thinking, but very elegant way to go about it.

**Comment by pitrou** [comment]
I didn't know https://herbie.uwplse.org/ ... That seems rather interesting :-o

**Pull request merged**